framework/include/base/Assembly.h

```diff
@@ -672,6 +672,43 @@ class Assembly
 
   /// Temporary work vector to keep from reallocating it
   std::vector<dof_id_type> _temp_dof_indices;
+
+public:
+  /**
+   * We can't add NodalBC values to the Jacobian (or preconditioning)
+   * matrix at the time they are computed -- we instead need to
+   * overwrite an entire row of values with the Jacobian associated to
+   * the NodalBC, which may be coupled to other variables, *after*
+   * matrix assembly is finished.  Hence, the Assembly object
+   * maintains a reference to this data which NodalBCs can write to,
+   * and which can be used later to properly set NodalBC jacobians.
+   * We use numeric_index_type for the index arrays (rather than
+   * dof_id_type) since that is what the SparseMatrix interface uses,
+   * but at the time of this writing, those two types are equivalent.
+   */
+  class CachedNodalBCData
+  {
+  public:
+    // Clears the cached data arrays
+    void clear()
+    {
+      cached_nodal_bc_vals.clear();
+      cached_nodal_bc_rows.clear();
+      cached_nodal_bc_cols.clear();
+    }
+
+    std::vector<Real> cached_nodal_bc_vals;
+    std::vector<numeric_index_type> cached_nodal_bc_rows;
+    std::vector<numeric_index_type> cached_nodal_bc_cols;
+  };
+
+  /**
+   * Return a writeable reference to the cached Nodal BC data.
+   */
+  CachedNodalBCData & cachedNodalBCData() { return _cached_nodal_bc_data; }
```

Comment:
IMO, there are 2 ways here:
  1. make the member variables in the class protected and add `CachedNodalBCData::addValue(i, j, value)`
  2. add `Assembly::cacheNodalBCValue(i, j, value)` and let the method do the caching.
   Change `CachedNodalBCData::clear()` to `Assembly::clearCachedNodalBCValues()`.
   How it is handled inside `Assembly` is no business of the caller and the caller does not have to get the writable reference. Unless, of course, there is a future plan that would require an access to `CachedNodalBCData`.

I'd go with (2).